PROD4POD-1426 - Merge upload workflows

fhd opened the pull request:
I've been experimenting with this on the side today, it seems a little cleaner to have a single workflow where the user can chose the upload targets - means we have less redundant logic in the separate workflows, and probably also means it's faster, since we e.g. only need to build/test "API and features" once.

JJ commented on .github/workflows/upload_all.yml:
```diff
@@ -1,26 +1,132 @@
+        run: |
+          if [[ ${{ github.event.inputs.targets }} =~ all|google-play ]]; then
+              echo "Uploading to Google Play"
+              echo "POLYPOD_UPLOAD_GOOGLE_PLAY=true" >> $GITHUB_ENV
```
OK, I see what you're doing here, turning a regex match (which GHA does not have) into a Boolean environment variable (which you can use either in step conditional expressions or in shell).

JJ replied:
OK, I see what you're doing here, turning a regex match (which GHA does not have) into a Boolean environment variable (which you can use either in step conditional expressions or in shell).

JJ commented on .github/workflows/upload_all.yml:
```diff
@@ -1,26 +1,132 @@
+        working-directory: platform/android
+        run: ./gradlew assemble -Psign=true
+        env:
+          POLYPOD_ANDROID_UPLOAD_KEYSTORE_FILE: keystore.tmp
```
This filename is repeated. Maybe create an env variable with the name at the job level?

JJ replied:
This filename is repeated. Maybe create an env variable with the name at the job level?

JJ commented on .github/workflows/upload_all.yml:
```diff
@@ -1,26 +1,132 @@
+        working-directory: platform/android
+      - name: Android - Set VERSION_NAME environment variable.
+        if: env.POLYPOD_UPLOAD_GOOGLE_PLAY == 'true'
+        run: echo "VERSION_NAME=$( ./getPackageInfo.sh /usr/local/lib/android/sdk/build-tools/30.0.0/aapt ./app/build/outputs/apk/release/app-release.apk versionName )" >> $GITHUB_ENV
```
Also repeated from above. Either turn the whole thing into a shell variable, or at least define the version string somewhere else.

JJ commented on .github/workflows/upload_all.yml:
```diff
@@ -1,26 +1,132 @@
+        working-directory: platform/android
+      - name: Android - Set VERSION_NAME environment variable.
+        if: env.POLYPOD_UPLOAD_GOOGLE_PLAY == 'true'
+        run: echo "VERSION_NAME=$( ./getPackageInfo.sh /usr/local/lib/android/sdk/build-tools/30.0.0/aapt ./app/build/outputs/apk/release/app-release.apk versionName )" >> $GITHUB_ENV
```
Also repeated from above. Either turn the whole thing into a shell variable, or at least define the version string somewhere else.

fhd replied:
This is just the Android workflow the way it was - I was actually interested in merging it into fewer steps, but didn't dare yet :D

fhd replied:
All these micro steps started bothering me, so I merged them into a single one now :D

JJ commented on .github/workflows/upload_all.yml:
```diff
@@ -1,26 +1,132 @@
+        id: vars
+        shell: bash
+        run: |
+          echo "##[set-output name=branch;]$(echo ${GITHUB_REF#refs/heads/})"
```
Clever :-)

fhd replied:
I think Tom wrote this, it's just from the Android workflow :D

JJ commented on .github/workflows/upload_all.yml:
```diff
@@ -1,26 +1,132 @@
+        id: vars
+        shell: bash
+        run: |
+          echo "##[set-output name=branch;]$(echo ${GITHUB_REF#refs/heads/})"
```
Clever :-)

JJ commented on .github/workflows/upload_all.yml:
```diff
@@ -1,26 +1,132 @@
+      - name: iOS - Install distribution prerequisites
+        if: env.POLYPOD_UPLOAD_APP_STORE == 'true'
+        working-directory: platform/ios
+        run: make install_distribution_prerequisites
```
`run` is usually the last key, but it does not really matter.

JJ commented on .github/workflows/upload_all.yml:
```diff
@@ -1,26 +1,132 @@
+      - name: iOS - Install distribution prerequisites
+        if: env.POLYPOD_UPLOAD_APP_STORE == 'true'
+        working-directory: platform/ios
+        run: make install_distribution_prerequisites
```
`run` is usually the last key, but it does not really matter.

fhd replied:
Moved it down everywhere! I <3 consistency.

JJ reviewed (approved):
Looking good overall. Just nits.

fhd commented on .github/workflows/upload_all.yml:
```diff
@@ -1,26 +1,132 @@
-      POLYPOD_ANDROID_STORE_SERVICE_ACCOUNT_JSON: ${{ secrets.POLYPOD_ANDROID_STORE_SERVICE_ACCOUNT_JSON }}
+  build_test_upload:
+    name: Build, test, and upload
+    runs-on: macos-11
```
Fair enough, I just copied this from how it was in the iOS workflow :D

fhd replied:
Changed it to _macos-latest_!

fhd commented on .github/workflows/upload_all.yml:
```diff
@@ -1,26 +1,132 @@
+    runs-on: macos-11
+    steps:
+      - name: Fail if uploading from non-release branch
+        if: |
```
Just took this over (in spirit) from the Android workflow. I did verify that this produces a failure as expected.

I don't think the branch filtering available for e.g. pushes works for `workflow_dispatch`, does it? There the user basically selects the branch to build on, and we need to somehow restrict the list of branches they can use, this has been our best idea so far I guess? But like I said, I just kept this working the way it was.

JJ replied:
As long as it's available as a variable, it's happening in that branch, so you should be able to filter. Let me check anyway.

JJ replied:
OK, no, you can't, but there are other options
* Make the user select the branch by defining it as an input
* Filtering at the job level, instead of the step level.

This is OK anyway, if you want it that way.

fhd replied:
I think if we specify it as an input, the user basically has to select it **twice**. I think this approach can survive this refactoring, given how it's the best I can think of. It's not a big deal to move to a better approach later if we find one.

fhd commented on .github/workflows/upload_all.yml:
```diff
@@ -1,26 +1,132 @@
+      - run: echo "Uploading ${{ github.ref }}"
+      - name: Set upload targets
+        run: |
+          if [[ ${{ github.event.inputs.targets }} =~ all|google-play ]]; then
```
I tried it with single quotes before, that wasn't valid syntax apparently - but still, good point, I'll see if I can find a way to quote it that makes bash happy

JJ replied:
Maybe the github action substitution engine got confused with them.

fhd replied:
Done! Turns out the thing left of the comparison operator can be a quoted string, but the expression to the right of the operator explicitly not - cause it's a regex :D Made it robust against whitespaces now. Also the matching was a bit error prone before

JJ reviewed (commented):
Answering comments.

fhd commented:
@JJ Thanks for having a look! What do you think in general - is this **better** than what we have, i.e. three separate workflows and using `workflow_call`? I'm still a bit undecided... My take right now:

Pros:
1. Using less GA minutes, since we're not building "API and features" twice (OTOH, we're running the Android build/tests/upload on the way more expensive macOS runners...)
2. Less duplication (mainly branch check, building API and features, and secrets names) - less code in general
3. Less likely to have partial uploads due to runner-specific problems - since everything runs on the same runner

Cons:
1. Bit more monolith spaghetti - I had to add comments to group steps in some way, because apparently there just is no way to group steps in GitHub actions together, `if` needs to be repeated on each step if you want to branch out, it's quite terrible in different ways
2. Should actually be **slower** overall, since the previous solution was to run the Android and iOS build in parallel

So, yeah, a bit torn... What do you think?

fhd commented:
@tstec-polypoly Also curious if you have an opinion - anyone with GA experience really :D

JJ commented:
> @JJ Thanks for having a look! What do you think in general - is this **better** than what we have, i.e. three separate workflows and using `workflow_call`? I'm still a bit undecided... My take right now:
> 
> Pros:
> 
>     1. Using less GA minutes, since we're not building "API and features" twice (OTOH, we're running the Android build/tests/upload on the way more expensive macOS runners...)
> 
>     2. Less duplication (mainly branch check, building API and features, and secrets names) - less code in general
> 
>     3. Less likely to have partial uploads due to runner-specific problems - since everything runs on the same runner
> 
> 
> Cons:
> 
>     1. Bit more monolith spaghetti - I had to add comments to group steps in some way, because apparently there just is no way to group steps in GitHub actions together, `if` needs to be repeated on each step if you want to branch out, it's quite terrible in different ways
> 
>     2. Should actually be **slower** overall, since the previous solution was to run the Android and iOS build in parallel
> 
> 
> So, yeah, a bit torn... What do you think?

As said above, I'm good with it. I think it's an improvement, and you can even create a top-level npm script target that calls gh on it, so it will be totally automated. It's not a test workflow, so it's not a big deal that it's slow. Don't mind monoliths, really.

fhd commented on .github/workflows/upload_all.yml:
```diff
@@ -1,26 +1,132 @@
+          node-version: 16
+          cache: "npm"
+          cache-dependency-path: "**/package-lock.json"
+      - name: API and features - Lint
```
Well, we have no guarantee here that tests and linting has already passed for whatever has most recently landed in the branch we're building. One of the areas where I find GA leaving a **lot** to be desired, but hey. We _could_ skip linting and tests though - then also on iOS/Android - assuming whoever runs this workflow knows what they're doing. Think that'd be better?

JJ replied:
linting is run whenever you push to any branch... Also tests. So I think they can safely be skipped...

fhd commented:
OK, so let's aim to land this then! Tried to improve it a bit more by addressing your comments @JJ, but might need to do a few more rounds of trial and error until it actually works again :D

tstec-polypoly commented on .github/workflows/upload_all.yml:
```diff
@@ -1,26 +1,129 @@
+          POLYPOD_ANDROID_UPLOAD_KEY_ALIAS: ${{ secrets.POLYPOD_ANDROID_UPLOAD_KEY_ALIAS }}
+        run: ./gradlew assemble -Psign=true
+
+      - name: Android - Determine release name
+        if: env.POLYPOD_UPLOAD_GOOGLE_PLAY == 'true'
+        run: |
+          version_code=$(./getPackageInfo.sh /usr/local/lib/android/sdk/build-tools/30.0.0/aapt ./app/build/outputs/apk/release/app-release.apk versionCode)
+          version_name=$(./getPackageInfo.sh /usr/local/lib/android/sdk/build-tools/30.0.0/aapt ./app/build/outputs/apk/release/app-release.apk versionName)
+          branch=${GITHUB_REF#refs/heads/}
+          sha_short=${GITHUB_SHA::7}
+          echo "##[set-output name=release_name;]$(echo $version_code ($version_name) - ${branch}@${sha_short})"
```
I'd forgotten about this funkiness, but seems it still works for setting steps.vars.outputs.release_name.

tstec-polypoly reviewed (approved):
LGTM for now. Chaining jobs would be nice-to-have.

fhd commented on .github/workflows/upload_all.yml:
```diff
@@ -1,26 +1,129 @@
+
+      - name: Android - Upload
+        # uses: r0adkll/upload-google-play@9745ef904e395471bca5696056a6ce8a60d18cf8
+        uses: polypoly-eu/upload-google-play@master
```
For now, yes, but if we add some what's new generation we can move to automatic deployment.

fhd commented on .github/workflows/upload_all.yml:
```diff
@@ -1,26 +1,129 @@
-      POLYPOD_APPLE_IOS_APPSTORE_API_PRIVATE_KEY_BASE64: ${{ secrets.POLYPOD_APPLE_IOS_APPSTORE_API_PRIVATE_KEY_BASE64 }}
-      POLYPOD_APPLE_IOS_APPSTORE_API_KEY_ID: ${{ secrets.POLYPOD_APPLE_IOS_APPSTORE_API_KEY_ID }}
-      POLYPOD_APPLE_IOS_APPSTORE_API_ISSUER_ID: ${{ secrets.POLYPOD_APPLE_IOS_APPSTORE_API_ISSUER_ID }}
+  build_test_upload:
```
I'm actually looking into this a bit right now - I don't think it'll be faster in any way, but I do think it'll fail quicker

JJ replied:
the problem is that, between jobs, there are limited options to pass information. Either you use output variables, or keyed caches. Using artifacts is considerably more difficult, and hacky.

fhd replied:
Yeah, I experimented with it a bit, and if that's OK by everyone, I would skip it for now. To do this efficiently, I would have to basically turn this workflow into two jobs that run in parallel, which are pretty much exactly the old iOS and Play Store upload jobs - i.e. not only would we have to redundantly build "API and features" -> Even the _code_ for that would have to be redundant.

Given how we're invoking this workflow at most once per day (in practice more like once a week), I think this is a better starting point than how it was.

JJ commented on .github/workflows/upload_all.yml:
```diff
@@ -1,26 +1,129 @@
-      POLYPOD_APPLE_IOS_APPSTORE_API_ISSUER_ID: ${{ secrets.POLYPOD_APPLE_IOS_APPSTORE_API_ISSUER_ID }}
+  build_test_upload:
+    name: Build, test, and upload
+    runs-on: macos-latest
```
Well, it's free for open source repos. The problem is that it's much slower, and quite unreliable.

JJ commented on platform/android/app/src/test/kotlin/coop/polypoly/polypod/updatenotification/UpdateNotificationTest.kt:
```diff
@@ -12,6 +13,7 @@ import org.robolectric.annotation.LooperMode
 @LooperMode(LooperMode.Mode.PAUSED)
 @RunWith(AndroidJUnit4::class)
 @Config(sdk = [Config.OLDEST_SDK])
+@Ignore("Not working on GitHub Actions")
```
Where did it fail? I can't find it...

fhd replied:
Was a flapper, removed it again in the mean time.

JJ commented on .github/workflows/upload_all.yml:
```diff
@@ -1,26 +1,133 @@
+          releaseName: ${{ steps.vars.outputs.release_name }}
+          releaseFiles: ./platform/android/app/build/outputs/apk/release/app-release.apk
+          track: internal
+          # whatsNewDirectory: distribution/whatsnew
```
Just delete this if you're not using it.

fhd replied:
Left it from the original workflow - guess we could indeed, but then again we're on the verge of adding non-draft publishing back, so if you don't mind too much I'll leave it for now, to finally merge this PR without having to do any further testing :D

JJ reviewed (approved):
Still looking fine!
